tools/README.md

```diff
@@ -0,0 +1,391 @@
+
+```
+
+### Add unit tests for the tool
+
+Envoy tooling is tested with `pytest`.
+
+The test runner expects a test file, in this case `tools/sometools/tests/test_mytool.py`.
+
+First, create the required directory if it is not present.
+
+```console
+$ mkdir tools/sometools/tests
+```
+
+Now add the following content to the `tools/sometools/tests/test_mytool.py` file:
+
+```python
+from unittest.mock import patch
+
+from tools.sometools import mytool
+
+
+def test_mytool_main():
+    with patch("tools.sometools.mytool.requests.get") as m_get:
+        with patch("tools.sometools.mytool.yaml.dump") as m_yaml:
+            with patch("tools.sometools.mytool.sys.stdout.write") as m_stdout:
+                assert mytool.main("PACKAGENAME") == 0
+    assert (
+        list(m_get.call_args)
+        == [('https://pypi.python.org/pypi/PACKAGENAME/json',), {}])
+    assert (
+        list(m_get.return_value.json.call_args)
+        == [(), {}])
+    assert (
+        list(m_yaml.call_args)
+        == [(m_get.return_value.json.return_value,), {}])
+    assert (
+        list(m_stdout.call_args)
+        == [(m_yaml.return_value,), {}])
+```
+
+This example use the mock library to patch all of the method calls, and
+then tests that they have been called with the expected values.
+
+You can run the test as follows:
+
+```console
+$ bazel run //tools/sometools:pytest_mytool
+INFO: Analyzed target //tools/sometools:pytest_mytool (0 packages loaded, 0 targets configured).
+INFO: Found 1 target...
+Target //tools/sometools:pytest_mytool up-to-date:
+  bazel-bin/tools/sometools/pytest_mytool
+INFO: Elapsed time: 0.247s, Critical Path: 0.07s
+INFO: 1 process: 1 internal.
+INFO: Build completed successfully, 1 total action
+================================ test session starts ===========================
+platform linux -- python 3.8.1, pytest-6.2.3, py-1.10.0, pluggy-0.13.1 -- /usr/bin/python3
+cachedir: .pytest_cache
+rootdir: /root/.cache/bazel/_bazel_root/f704bab1b165ed1368cb88f9f49e7532/execroot/envoy/bazel-out/k8-fastbuild/bin/tools/sometools/pytest_mytool.runfiles/envoy, configfile: pytest.ini
+plugins: cov-2.11.1
+collected 1 item
+
+tools/sometools/tests/test_mytool.py::test_mytool_main PASSED                                                                                                                             [100%]
+
+----------- coverage: platform linux, python 3.8.1-final-0 -----------
+Name                                                        Stmts   Miss  Cover
+-------------------------------------------------------------------------------
+/src/workspace/envoy/tools/sometools/mytool.py                  7      0   100%
+/src/workspace/envoy/tools/sometools/pytest_mytool.py           4      4     0%
+/src/workspace/envoy/tools/sometools/tests/test_mytool.py      11      0   100%
+-------------------------------------------------------------------------------
+TOTAL                                                          22      4    82%
```

Comment:
Can we enforce in CI that all tests that aren't otherwise tagged are required to ahve a minimum coverage percentage? This has been very successful in Envoy in driving very high coverage.

Reply:
yep - we can add that as we go

Reply:
(as an aside - it wouldnt be added here - it would need to be added at the end once all the tests had run - this code is 100% - just not by the end of this test specifically)

Reply:
ive set it to 95 - we can change as we need/want